src/qibosoq/qick_programs.py

```diff
@@ -174,42 +152,78 @@ def add_pulse_to_register(self, pulse: Pulse):
+
+        last_pulse_registered = {}
+        for idx in self.gen_chs:
+            last_pulse_registered[idx] = None
+
         for pulse in self.sequence:
             # time follows tproc CLK
-            time = self.soc.us2cycles(pulse.start * NS_TO_US)
+            time = self.soc.us2cycles(pulse.start)
 
-            qd_ch = self.qubits[pulse.qubit].drive.ports[0][1]
-            adc_ch = self.qubits[pulse.qubit].feedback.ports[0][1]
-            ro_ch = self.qubits[pulse.qubit].readout.ports[0][1]
-            gen_ch = qd_ch if pulse.type is PulseType.DRIVE else ro_ch
+            adc_ch = pulse.adc
+            gen_ch = pulse.dac
 
-            if not self.pulses_registered:
-                self.add_pulse_to_register(pulse)
-
-            if pulse.type is PulseType.DRIVE:
+            if pulse.type == "drive":
+                if not self.pulses_registered:
+                    self.add_pulse_to_register(pulse)
+                    last_pulse_registered[gen_ch] = pulse
                 self.pulse(ch=gen_ch, t=time)
-            elif pulse.type is PulseType.READOUT:
+            elif pulse.type == "readout":
+                if self.is_mux:
+                    if pulse not in muxed_pulses_executed:
+                        start = list(self.multi_ro_pulses)[len(muxed_ro_executed_time)]
+                        time = self.soc.us2cycles(start)
+                        self.add_muxed_readout_to_register(self.multi_ro_pulses[start])
+                        muxed_ro_executed_time.append(start)
+                        adcs = []
+                        for ro_pulse in self.multi_ro_pulses[start]:
+                            adcs.append(ro_pulse.adc)
+                            muxed_pulses_executed.append(ro_pulse)
+                    else:
+                        continue
+                else:
+                    if not self.pulses_registered:
+                        self.add_pulse_to_register(pulse)
+                    adcs = [adc_ch]
+
                 self.measure(
                     pulse_ch=gen_ch,
-                    adcs=[adc_ch],
+                    adcs=adcs,
                     adc_trig_offset=time + self.adc_trig_offset,
                     t=time,
                     wait=False,
                     syncdelay=self.syncdelay,
                 )
         self.wait_all()
-        self.sync_all(self.relax_delay)
+        if wait:
+            self.sync_all(self.relax_delay)
+
+    def is_pulse_equal(self, pulse_a: Pulse, pulse_b: Pulse) -> bool:
+        """Check if two pulses are equal, does not check the start time"""
+        if pulse_a is None or pulse_b is None:
+            return False
+        return (
+            pulse_a.frequency == pulse_b.frequency
+            and pulse_a.amplitude == pulse_b.amplitude
+            and pulse_a.relative_phase == pulse_b.relative_phase
+            and pulse_a.duration == pulse_b.duration
+            and pulse_a.type == pulse_b.type
+        )
```

Comment:
This function can also be defined as the `Pulse.__eq__` instead of putting it in the `BaseProgram`. Note that as it is now it is not using the program (`self`) anyway. 

I also cannot find where it is being used. If it's not maybe it could even be dropped, otherwise, you can replace it by `pulse_a == pulse_b` after defining it as `__eq__`.